### ISSUE TITLE
cabal haddock fails for foreign libraries

### ISSUE THREAD

**Issue by abooij** [haskell/cabal#4215]
When building the docs for the package in `cabal-testsuite/PackageTests/ForeignLibs`:

```
$ cabal haddock
Resolving dependencies...
Configuring my-foreign-lib-0.1.0.0...
Running Haddock for my-foreign-lib-0.1.0.0...
Preprocessing library my-foreign-lib-0.1.0.0...
Haddock coverage:
   0% (  0 /  2) in 'MyForeignLib.AnotherVal'
  Missing documentation for:
    Module header
    anotherVal (src/MyForeignLib/AnotherVal.hs:3)
Documentation created: dist/doc/html/my-foreign-lib/index.html
Preprocessing foreign library 'myforeignlib' for my-foreign-lib-0.1.0.0...
ppHsc2hs: broken closure: []
CallStack (from HasCallStack):
error, called at ./Distribution/Simple/PreProcess.hs:457:17 in
Cabal-1.25.0.0-LJMdZv6ScFWHI53bE5EyDD:Distribution.Simple.PreProcess
```

**Comment by abooij** [comment]
`cabal haddock`  succeeds for the ForeignLibs test project if the `build-depends` of the foreign libraries on `my-foreign-lib` (the package itself) is removed.
Further, from the docs for `dependencyClosure` (which ends up causing this error):

> Note that if the result is `Right []` it is because at least one of the original given `PackageId`s do not occur in the index.

In other words, in some situations, internal dependencies seem to not be resolved correctly.

**Comment by abooij** [comment]
In fact, this is the same problem for foreign-libraries as the similar one for executables, see #1085. The only difference is that the docs for foreign libraries are always built (this is a bug, presumably: the `--foreign-libraries` flag for `cabal haddock` is ignored - I'll patch this).

**Comment by abooij** [comment]
Actually, this bug is of the same type, but originates earlier: the preprocessing fails because of `ppHsc2hs`, so even if we correctly use `--foreign-libraries`, this bug is unavoidable as all components are preprocessed (even if no documentation is built). In the two bugs that I linked, hsc is not used, and so the bug is avoidable by not building documentation for those components.

**Comment by abooij** [comment]
`new-haddock` doesn't suffer from this bug.

**Comment by ezyang** [comment]
@abooij It sounds like you have done some great sleuthing on the problem. Would you be interested in fixing both the problem with foreign libraries and with executables? I think many people would be happy about this. Is it true that with any use of hsc with an internal library, haddock will fail?

**Comment by abooij** [comment]
@ezyang I would if I understood how dependencies are supposed to be resolved, and what is wrong with that in this instance. But I have no idea how to get started with that, as all these `BuildInfo` and `LocalBuildInfo` objects contain data that seems to simulate some things already being installed even though they're not, and it is not documented anywhere what data I am supposed to find where. So this smells like it needs a refactor, especially since new-haddock works correctly.

**Comment by abooij** [comment]
Oh and to answer @ezyang's last question:
Whenever you have a component that uses hsc, and this component depends on another component (in particular the `library`, say), haddock will fail.
However, you can avoid this particular bug with `ppHsc2hs` by first running `cabal build`, which does the preprocessing of the `.hsc` files correctly. Then haddock doesn't choke on the preprocessing, and builds the docs for the library component. But it still cannot build the docs for the component that uses hsc (such as the `foreign-library` or `executable`), as the bug of #1085 and #1919 is still present.

**Comment by ezyang** [comment]
@abooij The error, `broken closure: []`, is probably because the internal dep is not in the index that we're doing the closure in, right? This should not be hard to fix: we should not ONLY use `installedPkgs` from LocalBuildInfo, but also read in the internal package database (which will complete our package set.) Would you like to try this? That will fix #1085 and #1919, although you'll still have to build first (so that we get the inplace registrations) before haddocking.

If we want to avoid the build, we could crib logic from Distribution.Simple.Build:

```
  (\f -> foldM_ f (installedPkgs lbi) componentsToBuild) $ \index target -> do
    let comp = targetComponent target
        clbi = targetCLBI target
    initialBuildSteps distPref pkg_descr lbi clbi verbosity
    let bi     = componentBuildInfo comp
        progs' = addInternalBuildTools pkg_descr lbi bi (withPrograms lbi)
        lbi'   = lbi {
                   withPrograms  = progs',
                   withPackageDB = withPackageDB lbi ++ [internalPackageDB],
                   installedPkgs = index
                 }
    mb_ipi <- buildComponent verbosity (buildNumJobs flags) pkg_descr
                   lbi' suffixes comp clbi distPref
    return (maybe index (Index.insert `flip` index) mb_ipi)
```

Basically, we should compute the IPI for each component as we Haddock them, and then add them into the index as we go. That way the closure will work.

And YES, new-haddock does NOT have this problem because if you get to do per-component builds, the database will always have everything you need without any extra adding. Yay!
